src/Query/EagerLoader/Extractor.php

```diff
@@ -0,0 +1,119 @@
+<?php
+namespace Mongolid\Query\EagerLoader;
+
+use MongoDB\BSON\ObjectId;
+use Mongolid\Query\EagerLoader\Exception\EagerLoaderException;
+use Mongolid\Query\EagerLoader\Exception\InvalidModelKeyException;
+
+/**
+ * Responsible for extract ids from the model based on its
+ * relationship. You can use dot notations in order to
+ * extract ids from your embedded models.
+ */
+class Extractor
+{
+    /**
+     * This array will handle all related models extracted
+     * from the models passed by.
+     *
+     * @var array
+     */
+    private $relatedModels;
+
+    public function __construct(array $relatedModels)
+    {
+        $this->relatedModels = $relatedModels;
+    }
+
+    /**
+     * The model received here may be either an array from mongodb
+     * or a cached model as a serialized array.
+     * That's why we always will force an array to be used here.
+     * It will ensure that we are only working with model arrays.
+     */
+    public function extractFrom(array $model): array
+    {
+        foreach ($this->relatedModels as $eagerLoadKey => $loadModel) {
+            $key = $this->extractKeyFrom($loadModel);
+
+            // If the key extract contains '.' notations means that
+            // we want to extract models from embedded relations.
+            if ($this->keyHasDots($key)) {
+                $this->extractFromEmbeddedModel($eagerLoadKey, $model, $key);
+
+                continue;
+            }
+
+            $this->extractFromModel($eagerLoadKey, $model, $key);
+        }
+
+        return $this->relatedModels;
+    }
+
+    public function getRelatedModels(): array
+    {
+        return $this->relatedModels;
+    }
+
+    private function addIdFor(string $eagerLoadKey, $id): void
+    {
+        $this->relatedModels[$eagerLoadKey]['ids'][$id] = $id;
+    }
+
+    /**
+     * The user can use dot notations to get models from an
+     * embedded relationship. For now, we are only getting
+     * the first children for performance reasons.
+     *
+     * @example 'skus.shop_id' => I want all shop ids from embedded skus.
+     */
+    private function keyHasDots($key): bool
+    {
+        return str_contains($key, '.');
+    }
+
+    /**
+     * Using dot notations, the user can specify what ids on embedded
+     * models he wants to extract. So, our job is to loop on every
+     * embedded model to get the id specified on the key.
+     */
+    public function extractFromEmbeddedModel(string $eagerLoadKey, array $model, string $key): array
```

Comment:
Verify if this method really should be `public`. And if so, put in right place, before all `private` methods.